docs/shared/extending-nx/project-inference-plugins.md

```diff
@@ -0,0 +1,100 @@
+│  │  ├─ my-other-project-file
+│  │  ├─ project.json
+├─ nx.json
+├─ package.json
+
+```
+
+During initialization, we would expect to see "libs/my-project/my-other-project-file", "libs/nx-project/my-other-project-file", "libs/nx-project/project.json" all logged out to the console. Since `workspace.json` is not present, Nx was able to infer `my-project` from the layout.
+
+If `workspace.json` was present, `my-project` wouldn't be able to be inferred, since full project inference requires `workspace.json` to be absent.
+
+## Implementing a Project Target Configurator
+
+A project target configurator is a function that takes a path to a project file, and returns the targets inferred from that file.
+
+Plugins should export a function named `registerProjectTargets` that infers the targets from each matching project file. This function receives the path to the project file as its sole parameter.
+
+The `registerProjectTargets` function should return a `Record<string, TargetConfiguration>`, which describes the targets inferred for that specific project file.
+
+```typescript
+import {
+  TargetConfiguration
+} from '@nrwl/devkit';
+
+export function registerProjectTargets(
+  projectFilePath: string
+): Record<string, TargetConfiguration> {
+  return {
+    "build" {
+      /**
+       * This object should look exactly like a target
+       * configured inside `project.json`
+       */
+    }
+  }
+}
+```
+
+## Multiple Matches
+
+It is possible that the registerProjectTargets function may be called multiple times for one project. This could occur in a few cases, one of which is demonstrated above.
+
+- One plugin may list multiple file patterns, and a project may match more than one of them.
+- Multiple plugins may list similar patterns, and pick up the project separately.
+
+In the first case, the plugin that you are writing will be called into multiple times. If you return the same target (e.g. `build`) on each call, whichever is ran last would be the target that Nx calls into. The order that the function would be called is **NOT** guaranteed, so you should try to avoid this when possible. If specifying multiple patterns, they should either be mutually exclusive (e.g. one match per project) or the plugin should conditionally add targets based on the file passed in.
+
+In the second case, different plugins may attempt to register the same target on a project. If this occurs, whichever target was registered by the plugin listed latest in `nx.json` would be the one called into by Nx. As an example, assume `plugin-a`, `plugin-b`, and `plugin-c` all match a file and register `build` as a target. If `nx.json` included `"plugins": ["plugin-a", "plugin-b", "plugin-c"]`, running `nx build my-project` would run the target as defined by `"plugin-c"`. Alternatively, if `nx.json` included `"plugins": ["plugin-c", "plugin-b", "plugin-a"]`, running `nx build my-project` would run the target as defined by `"plugin-a"`
+
+## Visualizing the Project Graph
```

Comment:
I think a more fitting title is `Disabling Graph Cache During Development` or `Development Tips` or something similar. `Visualizing the Project Graph` refers to the first sentence of the entire paragraph. It's important, but from what I read the more important emphasis in this paragraph is that there's graph cache and it should be disabled during development.

Regarding `NX_DAEMON=false`, won't developers be able to see their logs in the deaemon log file?

Reply:
Yeah, that could be renamed. 

For `NX_DAEMON=false`, the quick answer is that yes, they could be seen inside the daemon log file. The slightly longer answer is that it still ends up being unexpected behavior, and the log file is a bit harder to find for some people, so the easy solution is to disable it during testing.